crates/bevy_render/src/camera/visible_entities.rs

```diff
@@ -197,14 +214,89 @@ mod rendering_mask_tests {
     }
 }
 
+//Computes effective visibility for entities.
+//
+//In hirerarchies the actual visiblity of an entity isn't only the current value,
+//but also depends on ancestors of the entity./ To avoid traversing each hierarchy
+//and determine the effective visibility for each entity, this system listens to
+//visiblity and hierarchy changes and only then computes a value to be cached and
+//used by other systems.
+pub fn visible_effective_system(
+    children_query: Query<&Children>,
+    changes_query: Query<
+        (Entity, Option<&Parent>),
+        (With<Visible>, Or<(Changed<Visible>, Changed<Parent>)>),
+    >,
+    mut visible_query: Query<(&Visible, Option<&mut VisibleEffective>)>,
+    mut commands: Commands,
+) {
+    fn update_effective(
+        entity: Entity,
+        is_visible_parent: bool,
+        children_query: &Query<&Children>,
+        mut visible_query: &mut Query<(&Visible, Option<&mut VisibleEffective>)>,
+        mut commands: &mut Commands,
+    ) {
+        if let Ok((visible, maybe_visible_effective)) = visible_query.get_mut(entity) {
+            let is_visible = visible.is_visible & is_visible_parent;
+            if let Some(mut visible_effective) = maybe_visible_effective {
+                visible_effective.is_transparent = visible.is_transparent;
+
+                if visible_effective.is_visible == is_visible {
+                    return;
+                }
+
+                visible_effective.is_visible = is_visible;
+            } else {
+                commands.entity(entity).insert(VisibleEffective {
+                    is_visible,
+                    is_transparent: visible.is_transparent,
+                });
+            }
+
+            if let Ok(children) = children_query.get(entity) {
+                for child in children.iter() {
+                    update_effective(
+                        *child,
+                        is_visible,
+                        &children_query,
+                        &mut visible_query,
+                        &mut commands,
+                    );
+                }
+            }
+        }
+    }
+
+    for (entity, maybe_parent) in changes_query.iter() {
+        if let Ok(is_visible) = match maybe_parent {
```

Comment:
Changes aren't guaranteed to iterate in hierarchy-order, which means that if we were to fix this impl to propagate "effective visibility" and only visit each node once, it might not be correct if we propagate in the wrong order.

Reply:
Exactly for that reason I do not update child entities which parent has no effective visibility. An advantage of the current implementation is that it completely avoids allocations to collect roots first. All spawned subtrees will be traversed only once.